torch/cuda/comm.py

```diff
@@ -74,22 +74,24 @@ def reduce_add(inputs, destination=None):
     """
     # TODO: try to find an input on another gpu, copy it,
     # and accumulate into the copy
+    if destination is None:
+        destination = torch.cuda.current_device()
     input_size = inputs[0].size()
     for i, inp in enumerate(inputs):
         assert inp.is_cuda, "reduce_add expects all inputs to be on GPUs"
+        if inp.get_device() == destination:
+            nccl_root = i
```

Comment:
The problem with this is that it's ambiguous if there are two tensors on destination device. Can you add `assert nccl_root is None` + `assert nccl_root is not None` to check that we really found the root after the loop?

Reply:
I can add asserts. nccl2 is actually removing support for the cases when a few tensors are on the same device, and I don't think you can have root that is not the device of one of the tensors.